components/tidb_query_codegen/src/rpn_function.rs

```diff
@@ -2098,6 +2126,117 @@ mod tests_normal {
+                ) -> tidb_query_common::Result<tidb_query_datatype::codec::data_type::VectorValue> {
+                    let arg = &self;
+                    let mut result = <Decimal as EvaluableRet>::ChunkedType::with_capacity(output_rows);
+                    use tidb_query_datatype::codec::data_type::{BitAndIterator, BitVec};
+                    let (vecs, fastpath, all_null) = {
+                        let mut vecs: Vec<&BitVec> = vec![];
+                        let mut fastpath = true;
+                        let mut all_null = false;
+                        let ((arg0, scalar_val), arg) = arg.get_bit_vec();
+                        if let Some(x) = arg0 {
+                            vecs.push(x);
+                            if !scalar_val {
+                                fastpath = false;
+                            }
+                        } else {
+                            if !scalar_val {
+                                all_null = true;
+                            }
+                        };
+                        if all_null {
+                            (vec![], false, true)
+                        } else if !fastpath {
+                            (vec![], false, false)
+                        } else {
+                            (vecs, true, false)
+                        }
+                    };
+                    if all_null {
+                        let mut result = <Decimal as EvaluableRet>::ChunkedType::with_capacity(output_rows);
+                        for i in 0..output_rows {
+                            result.push(None);
+                        }
+                        return Ok(Decimal::into_vector_value(result));
+                    }
+                    if !fastpath {
+                        for row_index in 0..output_rows {
+                            let (arg0, arg) = arg.extract(row_index);
+                            if arg0.is_none() {
+                                result.push(None);
+                                continue;
+                            }
+                            let arg0 = arg0.unwrap();
+                            result.push(foo(arg0)?);
+                        }
+                        return Ok(Decimal::into_vector_value(result));
+                    }
+                    for (row_index, val) in BitAndIterator::new(vecs.as_slice(), output_rows).enumerate() {
+                        if !val {
+                            result.push(None);
+                            continue;
+                        }
+                        let (arg0, arg) = arg.extract(row_index);
+                        let arg0 = arg0.unwrap();
+                        result.push(foo(arg0)?);
+                    }
+                    Ok(Decimal::into_vector_value(result))
+                }
+            }
+        };
+
+        // NOTE: Test via trim all space between code
+        assert_eq!(
+            expected.to_string().replace(" ", ""),
```

Comment:
In this case let mut a = 1; will equal to let muta = 1;. You can simply print the result by dbg!() and paste it here.

Reply:
But it's hard to read - -

Reply:
Can I call `rustfmt-nightly` here?

Reply:
Oh, I'd thought it would pretty print... I may mess up with other languages. Seems you could only tweak the spaces manually.. https://github.com/tikv/tikv/blob/9330514b93eea30fa3afb167a68251ffb520cb7b/components/tidb_query_codegen/src/rpn_function.rs#L2094-L2096

Reply:
I worked out a new way to compare token tree directly, PTAL.

Reply:
Awesome! That's a good approach.